sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,526 @@
+  message AfterAll {
+    repeated Trigger sub_trigger = 1;
+  }
+
+  message AfterAny {
+    repeated Trigger sub_trigger = 1;
+  }
+
+  message AfterEach {
+    repeated Trigger sub_trigger = 1;
+  }
+
+  message AfterEndOfWindow {
+    // Optional
+    Trigger early_firings = 1;
+
+    // Optional
+    Trigger late_firings = 2;
+  }
+
+  message AfterProcessingTimeDelay {
+    int64 delay_millis = 1;
+  }
+
+  message AfterSynchronizedProcessingTimeDelay {
+    int64 delay_millis = 1;
+  }
+
+  message Never { }
+
+  message Repeat {
+    Trigger sub_trigger = 1;
+  }
+
+  message OrFinally {
+    Trigger main = 1;
+    Trigger finally = 2;
+  }
+
+  message Default { }
+
+  oneof trigger {
+    AfterAll after_all = 1;
+    AfterAny after_any = 2;
+    AfterEach after_each = 3;
+    AfterEndOfWindow after_end_of_widow = 4;
+    AfterProcessingTimeDelay after_processing_time_delay = 5;
+    AfterSynchronizedProcessingTimeDelay after_synchronized_processing_time_delay = 6;
+    Never never = 7;
+    Repeat repeat = 8;
+    OrFinally or_finally = 9;
+    Default default = 10;
+  }
+}
+
+// A specification for how to "side input" a PCollection.
+message SideInput {
+
+  // (Required) URN of the access pattern required by the `view_fn` to present
+  // the desired SDK-specific interface to a UDF.
+  //
+  // This access pattern defines the SDK harness <-> Runner Harness RPC
+  // interface for accessing a side input.
+  //
+  // The only access pattern intended for Beam, because of its superior
+  // performance possibilities, is "urn:beam:sideinput:multimap" (or some such
+  // URN)
+  string access_pattern = 1;
+
+  // (Required) A UDF that adapts a particular access_pattern to a user-facing
+  // view type.
+  //
+  // For example, View.asSingleton() may include a `view_fn` that adapts a
+  // specially-designed multimap to a single value per window.
+  //
+  // TODO: this will actually be reused a lot; consider making it by reference
+  FunctionSpec view_fn = 2;
+
+  // (Required) A UDF that maps a main input window to a side input window.
+  //
+  // For example, when the main input is in fixed windows of one hour, this
+  // can specify that the side input should be accessed according to the day
+  // in which that hour falls.
+  FunctionSpec window_mapping_fn = 3;
+}
+
+// An environment for executing UDFs. Generally an SDK container URL, but
+// there can be many for a single SDK, for example to provide dependency
+// isolation.
+message Environment {
+
+  // (Required) A pipeline-scoped unique id for this node
+  int64 id = 1;
+
+  // (Required) The URL of a container
+  //
+  // TODO: reconcile with Fn API's DockerContainer structure by
+  // adding adequate metadata to know how to interpret the container
+  string url = 2;
+}
+
+// Description of a user-defined function, including the environment/SDK in
+// which to execute it
+message FunctionSpec {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
```

Comment:
Everything that embeds a FunctionSpec and need to be uniquely identified has an `id`. Do we really need the id or can it at least be an optional field?